File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/events/EventsViewModel.java
package com.eulersbridge.isegoria.feed.events;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Transformations;
import android.support.annotation.NonNull;

import com.eulersbridge.isegoria.IsegoriaApp;
import com.eulersbridge.isegoria.network.api.models.Event;
import com.eulersbridge.isegoria.util.data.SingleLiveData;
import com.eulersbridge.isegoria.util.data.RetrofitLiveData;

import java.util.List;

@SuppressWarnings("WeakerAccess")
public class EventsViewModel extends AndroidViewModel {

    private LiveData<List<Event>> eventsList;

    public EventsViewModel(@NonNull Application application) {
        super(application);
    }

    LiveData<List<Event>> getEvents() {
        IsegoriaApp isegoriaApp = getApplication();

        return Transformations.switchMap(isegoriaApp.loggedInUser, user -> {
            if (user != null && user.institutionId != null) {
                eventsList = new RetrofitLiveData<>(isegoriaApp.getAPI().getEvents(user.institutionId));
                return eventsList;
            }

            return new SingleLiveData<>(null);
        });
    }

    @Override
    protected void onCleared() {
        if (eventsList instanceof RetrofitLiveData)
            ((RetrofitLiveData) eventsList).cancel();
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/events/EventsViewModel.kt
package com.eulersbridge.isegoria.feed.events

import android.arch.lifecycle.MutableLiveData
import com.eulersbridge.isegoria.data.Repository
import com.eulersbridge.isegoria.network.api.model.Event
import com.eulersbridge.isegoria.util.BaseViewModel
import com.eulersbridge.isegoria.util.extension.subscribeSuccess
import com.eulersbridge.isegoria.util.extension.zipWithTimer
import javax.inject.Inject

class EventsViewModel @Inject constructor(private val repository: Repository) : BaseViewModel() {

    internal var isRefreshing = MutableLiveData<Boolean>()
    internal val events = MutableLiveData<List<Event>>()

    init {
        isRefreshing.value = false
        refresh()
    }

    internal fun refresh() {
        isRefreshing.postValue(true)

        zipWithTimer(repository.getEvents())
                .subscribeSuccess {
                    isRefreshing.postValue(false)
                    events.postValue(it)
                }.addToDisposable()
    }
}
